Library/Source/NativeEngine.cpp

```diff
@@ -753,81 +753,104 @@ namespace babylon
         m_engineState |= ALPHA_MODE[blendMode];
     }
 
-    void NativeEngine::SetMatrix(const Napi::CallbackInfo& info)
+    void NativeEngine::SetInt(const Napi::CallbackInfo& info)
+    {
+        const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
+        const auto value = info[1].As<Napi::Number>().FloatValue();
+        m_currentProgram->SetUniform(uniformData->Handle, gsl::make_span(&value, 1));
+    }
+
+    template<int size, typename arrayType> void NativeEngine::SetTypeArrayN(const Napi::CallbackInfo& info)
+    {
+        const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
+        const auto array = info[1].As<arrayType>();
+
+        size_t elementLength = array.ElementLength();
+
+        m_scratch.clear();
+        for (size_t index = 0; index < elementLength; index += size)
+        {
+            const float values[] = { static_cast<float>(array[index])
+                , (size > 1) ? static_cast<float>(array[index + 1]) : 0.f
+                , (size > 2) ? static_cast<float>(array[index + 2]) : 0.f
+                , (size > 3) ? static_cast<float>(array[index + 3]) : 0.f };
+            m_scratch.insert(m_scratch.end(), values, values + 4);
+        }
+
+        m_currentProgram->SetUniform(uniformData->Handle, m_scratch, elementLength / size);
+    }
+
+    template<int size> void NativeEngine::SetFloatN(const Napi::CallbackInfo& info)
+    {
+        const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
+        const float values[] = { info[1].As<Napi::Number>().FloatValue(),
+                                (size > 1) ? info[2].As<Napi::Number>().FloatValue() : 0.f,
+                                (size > 2) ? info[3].As<Napi::Number>().FloatValue() : 0.f,
+                                (size > 3) ? info[4].As<Napi::Number>().FloatValue() : 0.f };
+
+        m_currentProgram->SetUniform(uniformData->Handle, values);
+    }
+
+    template<int size> void NativeEngine::SetMatrixN(const Napi::CallbackInfo& info)
     {
         const auto uniformData = info[0].As<Napi::External<UniformInfo>>().Data();
         const auto matrix = info[1].As<Napi::Float32Array>();
 
         const size_t elementLength = matrix.ElementLength();
-        assert(elementLength == 16);
+        assert(elementLength == size * size);
+
+        std::array<float, 16> matrixValues{ };
+
+        size_t index = 0;
+        for (int line = 0; line < size; line++)
+        {
+            for (int col = 0; col < size; col++)
+            {
+                matrixValues[line * 4 + col] = matrix[index++];
+            }
+        }
```

Comment:
Is there a performance issue here when matrix is 4x4? We used to directly use the data, but now we copy the data with a double for loop? Maybe we should specialize the 4x4 case?